File: src/main/java/com/binance/api/client/impl/BinanceApiAsyncRestClientImpl.java
package com.binance.api.client.impl;

import com.binance.api.client.BinanceApiAsyncRestClient;
import com.binance.api.client.domain.account.*;
import com.binance.api.client.domain.account.request.AllOrdersRequest;
import com.binance.api.client.domain.account.request.CancelOrderRequest;
import com.binance.api.client.domain.account.request.OrderRequest;
import com.binance.api.client.domain.account.request.OrderStatusRequest;
import com.binance.api.client.domain.event.ListenKey;
import com.binance.api.client.domain.general.ExchangeInfo;
import com.binance.api.client.domain.general.ServerTime;
import com.binance.api.client.domain.market.*;
import com.binance.api.client.exception.BinanceApiException;
import retrofit2.Call;

import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

import static com.binance.api.client.constant.BinanceApiConstants.DEFAULT_RECEIVING_WINDOW;
import static com.binance.api.client.impl.BinanceApiServiceGenerator.createService;
import static java.lang.System.currentTimeMillis;

/**
 * Implementation of Binance's REST API using Retrofit with asynchronous/non-blocking method calls.
 */
public class BinanceApiAsyncRestClientImpl implements BinanceApiAsyncRestClient {

  public final BinanceApiService binanceApiService;

  public BinanceApiAsyncRestClientImpl(String apiKey, String secret) {
    binanceApiService = createService(BinanceApiService.class, apiKey, secret);
  }

  static <T> CompletableFuture<T> futureGlue(Call<T> call) {
    CompletableFuture<T> ret = new CompletableFuture<>();

    try {
      call.enqueue(new BinanceApiCallbackAdapter<T>(ret::complete) {
        @Override
        public void onFailure(Call call, Throwable throwable) {
          ret.completeExceptionally(throwable);
        }
      });
    } catch (BinanceApiException throwable) {
      ret.completeExceptionally(throwable);
    }
    return ret;
  }

  // General endpoints

  @Override
  public CompletableFuture<Void> ping() {
    return futureGlue(binanceApiService.ping());
  }

  @Override
  public CompletableFuture<ServerTime> getServerTime() {
    return futureGlue(binanceApiService.getServerTime());
  }

  @Override
  public CompletableFuture<ExchangeInfo> getExchangeInfo() {
    return futureGlue(binanceApiService.getExchangeInfo());
  }

  // Market Data endpoints

  @Override
  public CompletableFuture<OrderBook> getOrderBook(String symbol, Integer limit) {
    return futureGlue(binanceApiService.getOrderBook(symbol, limit));
  }

  @Override
  public CompletableFuture<List<AggTrade>> getAggTrades(String symbol, Optional<String> fromId, Optional<Integer> limit, Optional<Long> startTime, Optional<Long> endTime) {
    return futureGlue(binanceApiService.getAggTrades(
        symbol,
        fromId.orElse(null),
        limit.orElse(null),
        startTime.orElse(null),
        endTime.orElse(null)
    ));
  }

  @Override
  public CompletableFuture<List<Candlestick>> getCandlestickBars(String symbol, CandlestickInterval interval, Optional<Integer> limit, Optional<Long> startTime, Optional<Long> endTime) {
    return futureGlue(binanceApiService.getCandlestickBars(
        symbol,
        interval.getIntervalId(),
        limit.orElse(null),
        startTime.orElse(null),
        endTime.orElse(null)
    ));
  }

  @Override
  public CompletableFuture<TickerStatistics> get24HrPriceStatistics(String symbol) {
    return futureGlue(binanceApiService.get24HrPriceStatistics(symbol));
  }

  @Override
  public CompletableFuture<List<TickerPrice>> getAllPrices() {
    return futureGlue(binanceApiService.getLatestPrices());
  }

  @Override
  public CompletableFuture<List<BookTicker>> getBookTickers() {
    return futureGlue(binanceApiService.getBookTickers());
  }

  @Override
  public CompletableFuture<NewOrderResponse> newOrder(NewOrder order) {
    return futureGlue(binanceApiService.newOrder(
        order.symbol,
        order.side,
        order.type,
        order.timeInForce,
        order.quantity,
        order.price.orElse(null),
        order.stopPrice.orElse(null),
        order.icebergQty.orElse(null),
        order.recvWindow,
        order.timestamp
    ));
  }

  @Override
  public CompletableFuture<Void> newOrderTest(NewOrder in) {
    return futureGlue(binanceApiService.newOrderTest(
        in.symbol,
        in.side,
        in.type,
        in.timeInForce,
        in.quantity,
        in.price.orElse(null),
        in.stopPrice.orElse(null),
        in.icebergQty.orElse(null),
        in.recvWindow,
        in.timestamp
    ));
  }

  // Account endpoints

  @Override
  public CompletableFuture<Order> getOrderStatus(OrderStatusRequest in) {
    return futureGlue(binanceApiService.orderStatus(
        in.symbol,
        in.orderId.orElse(null),
        in.origClientOrderId.orElse(null),
        in.recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        in.timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<Void> cancelOrder(CancelOrderRequest in) {
    return futureGlue(binanceApiService.cancelOrder(
        in.symbol,
        in.orderId.orElse(null),
        in.origClientOrderId.orElse(null),
        in.newClientOrderId.orElse(null),
        in.recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        in.timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<List<Order>> getOpenOrders(OrderRequest in) {
    return futureGlue(binanceApiService.openOrders(
        in.symbol,
        in.recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        in.timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<List<Order>> getAllOrders(AllOrdersRequest in) {
    return futureGlue(binanceApiService.allOrders(
        in.symbol,
        in.orderId.orElse(null),
        in.limit.orElse(null),
        in.recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        in.timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<Account> getAccount(Optional<Long> recvWindow, Optional<Long> timestamp) {
    return futureGlue(binanceApiService.account(
        recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<List<Trade>> getMyTrades(String symbol, Optional<Integer> limit, Optional<Long> fromId, Optional<Long> recvWindow, Optional<Long> timestamp) {
    return futureGlue(
    binanceApiService.getMyTrades(
        symbol,
        limit.orElse(null),
        fromId.orElse(null),
        recvWindow.orElse(DEFAULT_RECEIVING_WINDOW),
        timestamp.orElse(currentTimeMillis())
    ));
  }

  @Override
  public CompletableFuture<Void> withdraw(String asset, String address, String amount, Optional<String> name) {
    return futureGlue(binanceApiService.withdraw(asset, address, amount, name.orElse(null), DEFAULT_RECEIVING_WINDOW, currentTimeMillis())
        );
  }

  @Override
  public CompletableFuture<DepositHistory> getDepositHistory(String asset) {
    return futureGlue(binanceApiService.depositHistory(asset, DEFAULT_RECEIVING_WINDOW, currentTimeMillis())
        );
  }

  @Override
  public CompletableFuture<WithdrawHistory> getWithdrawHistory(String asset) {
    return futureGlue(binanceApiService.withdrawHistory(asset, DEFAULT_RECEIVING_WINDOW, currentTimeMillis())
        );
  }

  @Override
  public CompletableFuture<DepositAddress> getDepositAddress(String asset) {
    return futureGlue(binanceApiService.depositAddress(asset, DEFAULT_RECEIVING_WINDOW, currentTimeMillis())
        );
  }

  // User stream endpoints

  @Override
  public CompletableFuture<ListenKey> startUserDataStream() {
    return futureGlue(binanceApiService.startUserDataStream());
  }

  @Override
  public CompletableFuture<Void> keepAliveUserDataStream(ListenKey listenKey) {
    return futureGlue(binanceApiService.keepAliveUserDataStream(listenKey.listenKey));
  }

  @Override
  public CompletableFuture<Void> closeUserDataStream(ListenKey listenKey) {
    return futureGlue(binanceApiService.closeAliveUserDataStream(listenKey.listenKey));
  }
}

File: src/main/java/com/binance/api/client/impl/BinanceApiAsyncRestClientImpl.scala
package com.binance.api.client.impl

import java.lang.System.currentTimeMillis

import com.binance.api.client.BinanceApiAsyncRestClient
import com.binance.api.client.constant.BinanceApiConstants.DEFAULT_RECEIVING_WINDOW
import com.binance.api.client.domain._
import com.binance.api.client.domain.account._
import com.binance.api.client.domain.account.request.{
  AllOrdersRequest,
  CancelOrderRequest,
  OrderRequest,
  OrderStatusRequest
}
import com.binance.api.client.domain.event.ListenKey
import com.binance.api.client.domain.general.{ExchangeInfo, ServerTime}
import com.binance.api.client.domain.market._
import com.binance.api.client.json.Decoders._

import scala.concurrent.{ExecutionContext, Future}

trait AsJava[From, To] {
  def convert(from: From): To
}

object AsJava {
  def apply[From, To](from: From)(implicit asJava: AsJava[From, To]): To =
    asJava.convert(from)

  def apply[From, To](fromOpt: Option[From])(implicit asJava: AsJava[From, To]): To =
    fromOpt match {
      case Some(from) => apply(from)
      case None       => null.asInstanceOf[To]
    }

  def or[From, To](fromOpt: Option[From], or: To)(implicit asJava: AsJava[From, To]): To =
    fromOpt match {
      case Some(from) => apply(from)
      case None       => or
    }

  implicit val x0: AsJava[String, String]          = t => t
  implicit val x1: AsJava[Long, java.lang.Long]    = t => t
  implicit val x2: AsJava[Int, java.lang.Integer]  = t => t
  implicit val x3: AsJava[Instant, java.lang.Long] = _.millis
  implicit val x4: AsJava[Asset, String]           = _.value
  implicit val x5: AsJava[Quantity, String]        = _.value.toString
  implicit val x6: AsJava[Price, String]           = _.value.toString
  implicit val x7: AsJava[OrderId, java.lang.Long] = _.value
  implicit val x8: AsJava[Symbol, String]          = _.value.toUpperCase

  implicit def instanceOpt[E <: java.lang.Enum[E]]: AsJava[E, E] = e => e
}

/**
  * Implementation of Binance's REST API using Retrofit with asynchronous/non-blocking method calls.
  */
class BinanceApiAsyncRestClientImpl(service: BinanceApiService)(implicit ex: ExecutionContext)
    extends BinanceApiAsyncRestClient {

  override def ping: Future[Unit] =
    RunRequest[Unit](service.ping)

  override def getServerTime: Future[ServerTime] =
    RunRequest[ServerTime](service.getServerTime)

  override def getExchangeInfo: Future[ExchangeInfo] =
    RunRequest[ExchangeInfo](service.getExchangeInfo)

  override def getOrderBook(symbol: Symbol, limit: Int): Future[OrderBook] =
    RunRequest[OrderBook](service.getOrderBook(symbol.value, limit))

  def getAggTrades(symbol:    Symbol,
                   fromId:    Option[String] = None,
                   limit:     Option[Int] = None,
                   startTime: Option[Instant] = None,
                   endTime:   Option[Instant] = None): Future[List[AggTrade]] =
    RunRequest[List[AggTrade]](
      service
        .getAggTrades(
          AsJava(symbol),
          AsJava(fromId),
          AsJava(limit),
          AsJava(startTime),
          AsJava(endTime)
        )
    )

  def getCandlestickBars(symbol:    Symbol,
                         interval:  CandlestickInterval,
                         limit:     Option[Int] = None,
                         startTime: Option[Instant] = None,
                         endTime:   Option[Instant] = None): Future[List[Candlestick]] =
    RunRequest[List[Candlestick]](
      service
        .getCandlestickBars(AsJava(symbol),
                            AsJava(interval.intervalId),
                            AsJava(limit),
                            AsJava(startTime),
                            AsJava(endTime))
    )

  override def get24HrPriceStatistics(symbol: Symbol): Future[TickerStatistics] =
    RunRequest[TickerStatistics](
      service.get24HrPriceStatistics(
        AsJava(symbol)
      )
    )

  override def getAllPrices: Future[List[TickerPrice]] =
    RunRequest[List[TickerPrice]](service.getLatestPrices)

  override def getBookTickers: Future[List[BookTicker]] =
    RunRequest[List[BookTicker]](service.getBookTickers)

  override def newOrder(order: NewOrder): Future[NewOrderResponse] =
    RunRequest[NewOrderResponse](
      service.newOrder(
        AsJava(order.symbol),
        AsJava(order.side),
        AsJava(order.`type`),
        AsJava(order.timeInForce),
        AsJava(order.quantity),
        AsJava(order.price),
        AsJava(order.stopPrice),
        AsJava(order.icebergQty),
        AsJava(order.recvWindow),
        AsJava(order.timestamp)
      )
    )

  override def newOrderTest(in: NewOrder): Future[Unit] =
    RunRequest[Unit](
      service.newOrderTest(
        AsJava(in.symbol),
        AsJava(in.side),
        AsJava(in.`type`),
        AsJava(in.timeInForce),
        AsJava(in.quantity),
        AsJava(in.price),
        AsJava(in.stopPrice),
        AsJava(in.icebergQty),
        AsJava(in.recvWindow),
        AsJava(in.timestamp)
      )
    )

  override def getOrderStatus(in: OrderStatusRequest): Future[Order] =
    RunRequest[Order](
      service.orderStatus(
        AsJava(in.symbol),
        AsJava(in.orderId),
        AsJava(in.origClientOrderId),
        AsJava.or(in.recvWindow, DEFAULT_RECEIVING_WINDOW),
        AsJava.or(in.timestamp, currentTimeMillis)
      )
    )

  override def cancelOrder(in: CancelOrderRequest): Future[Unit] =
    RunRequest[Unit](
      service.cancelOrder(
        AsJava(in.symbol),
        AsJava(in.orderId),
        AsJava(in.origClientOrderId),
        AsJava(in.newClientOrderId),
        AsJava.or(in.recvWindow, DEFAULT_RECEIVING_WINDOW),
        AsJava.or(in.timestamp, currentTimeMillis)
      )
    )

  override def getOpenOrders(in: OrderRequest): Future[List[Order]] =
    RunRequest[List[Order]](
      service.openOrders(AsJava(in.symbol),
                         AsJava.or(in.recvWindow, DEFAULT_RECEIVING_WINDOW),
                         AsJava.or(in.timestamp, currentTimeMillis))
    )

  override def getAllOrders(in: AllOrdersRequest): Future[List[Order]] =
    RunRequest[List[Order]](
      service.allOrders(
        AsJava(in.symbol),
        AsJava(in.orderId),
        AsJava(in.limit),
        AsJava.or(in.recvWindow, DEFAULT_RECEIVING_WINDOW),
        AsJava.or(in.timestamp, currentTimeMillis)
      )
    )

  def getAccount(recvWindow: Option[Long], timestamp: Option[Instant]): Future[Account] =
    RunRequest[Account](
      service.account(
        AsJava.or(recvWindow, DEFAULT_RECEIVING_WINDOW),
        AsJava.or(timestamp, currentTimeMillis)
      )
    )

  def getMyTrades(symbol:     Symbol,
                  limit:      Option[Int],
                  fromId:     Option[Long],
                  recvWindow: Option[Long],
                  timestamp:  Option[Instant]): Future[List[Trade]] =
    RunRequest[List[Trade]](
      service.getMyTrades(
        AsJava(symbol),
        AsJava(limit),
        AsJava(fromId),
        AsJava.or(recvWindow, DEFAULT_RECEIVING_WINDOW),
        AsJava.or(timestamp, currentTimeMillis)
      )
    )

  override def withdraw(asset: Asset, address: String, amount: String, name: Option[String] = None): Future[Unit] =
    RunRequest[Unit](
      service.withdraw(
        AsJava(asset),
        AsJava(address),
        AsJava(amount),
        AsJava(name),
        AsJava(DEFAULT_RECEIVING_WINDOW),
        AsJava(currentTimeMillis)
      )
    )

  override def getDepositHistory(asset: Asset): Future[DepositHistory] =
    RunRequest[DepositHistory](
      service
        .depositHistory(
          AsJava(asset),
          AsJava(DEFAULT_RECEIVING_WINDOW),
          AsJava(currentTimeMillis)
        )
    )

  override def getWithdrawHistory(asset: Asset): Future[WithdrawHistory] =
    RunRequest[WithdrawHistory](
      service
        .withdrawHistory(
          AsJava(asset.value),
          AsJava(DEFAULT_RECEIVING_WINDOW),
          AsJava(currentTimeMillis)
        )
    )

  override def getDepositAddress(asset: Asset): Future[DepositAddress] =
    RunRequest[DepositAddress](
      service
        .depositAddress(
          AsJava(asset),
          AsJava(DEFAULT_RECEIVING_WINDOW),
          AsJava(currentTimeMillis)
        )
    )

  override def startUserDataStream: Future[ListenKey] =
    RunRequest[ListenKey](service.startUserDataStream)

  override def keepAliveUserDataStream(listenKey: ListenKey): Future[Unit] =
    RunRequest[Unit](service.keepAliveUserDataStream(listenKey.listenKey))

  override def closeUserDataStream(listenKey: ListenKey): Future[Unit] =
    RunRequest[Unit](service.closeAliveUserDataStream(listenKey.listenKey))
}
